src/TimeoutMigrationTool/ASQ/ASQEndpointMigrator.cs

```diff
@@ -0,0 +1,111 @@
+﻿namespace Particular.TimeoutMigrationTool.ASQ
+{
+    using System;
+    using Microsoft.Azure.Cosmos.Table;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Cosmos.Table.Queryable;
+
+    public class ASQEndpointMigrator : ITimeoutsTarget.IEndpointTargetBatchMigrator
+    {
+        public ASQEndpointMigrator(CloudTableClient client, string destinationTimeoutTableName, string stagingTableName)
+        {
+            this.client = client;
+            this.destinationTimeoutTableName = destinationTimeoutTableName;
+            this.stagingTableName = stagingTableName;
+        }
+
+        public async ValueTask<int> StageBatch(IReadOnlyList<TimeoutData> timeouts, int batchNumber)
+        {
+            await PurgeStagingTable();
```

Comment:
As far as I remember purging is not efficient at all when there is large amounts of data in there. In most cases, the recommendation is to delete and recreate the table. There is one caveat do you have to deal with the delete being eventually consistent and when you try to create it right after you need to retry on 409 (conflict)

The doco states though deleting and recreating can take a while https://docs.microsoft.com/en-us/rest/api/storageservices/Delete-Table?redirectedfrom=MSDN#remarks but in practice I've never seen it taking so long. Btw I'm not against sticking to the current way. I think though we should align the approaches in the project because the other branch currently uses delete and recreate

Reply:
Laila knows everything